File: src/marktoflow/core/state.py
"""
State persistence for marktoflow framework.

Provides SQLite-based state storage for workflow checkpoints,
execution history, and recovery.
"""

from __future__ import annotations

import json
import sqlite3
import logging
from contextlib import contextmanager
from dataclasses import dataclass, asdict
from datetime import datetime, timedelta
from enum import Enum
from pathlib import Path
from typing import Any, Generator

logger = logging.getLogger(__name__)


class ExecutionStatus(Enum):
    """Status of a workflow execution."""

    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"
    PAUSED = "paused"


@dataclass
class ExecutionRecord:
    """Record of a workflow execution."""

    run_id: str
    workflow_id: str
    workflow_path: str
    status: ExecutionStatus
    started_at: datetime
    completed_at: datetime | None = None
    current_step: int = 0
    total_steps: int = 0
    agent: str | None = None
    inputs: dict[str, Any] | None = None
    outputs: dict[str, Any] | None = None
    error: str | None = None
    metadata: dict[str, Any] | None = None

    def to_dict(self) -> dict[str, Any]:
        """Convert to dictionary for storage."""
        data = asdict(self)
        data["status"] = self.status.value
        data["started_at"] = self.started_at.isoformat()
        data["completed_at"] = self.completed_at.isoformat() if self.completed_at else None
        data["inputs"] = json.dumps(self.inputs) if self.inputs else None
        data["outputs"] = json.dumps(self.outputs) if self.outputs else None
        data["metadata"] = json.dumps(self.metadata) if self.metadata else None
        return data

    @classmethod
    def from_row(cls, row: sqlite3.Row) -> "ExecutionRecord":
        """Create from database row."""
        return cls(
            run_id=row["run_id"],
            workflow_id=row["workflow_id"],
            workflow_path=row["workflow_path"],
            status=ExecutionStatus(row["status"]),
            started_at=datetime.fromisoformat(row["started_at"]),
            completed_at=datetime.fromisoformat(row["completed_at"])
            if row["completed_at"]
            else None,
            current_step=row["current_step"],
            total_steps=row["total_steps"],
            agent=row["agent"],
            inputs=json.loads(row["inputs"]) if row["inputs"] else None,
            outputs=json.loads(row["outputs"]) if row["outputs"] else None,
            error=row["error"],
            metadata=json.loads(row["metadata"]) if row["metadata"] else None,
        )


@dataclass
class StepCheckpoint:
    """Checkpoint for a workflow step."""

    run_id: str
    step_index: int
    step_name: str
    status: ExecutionStatus
    started_at: datetime
    completed_at: datetime | None = None
    inputs: dict[str, Any] | None = None
    outputs: dict[str, Any] | None = None
    error: str | None = None
    retry_count: int = 0

    def to_dict(self) -> dict[str, Any]:
        """Convert to dictionary for storage."""
        return {
            "run_id": self.run_id,
            "step_index": self.step_index,
            "step_name": self.step_name,
            "status": self.status.value,
            "started_at": self.started_at.isoformat(),
            "completed_at": self.completed_at.isoformat() if self.completed_at else None,
            "inputs": json.dumps(self.inputs) if self.inputs else None,
            "outputs": json.dumps(self.outputs) if self.outputs else None,
            "error": self.error,
            "retry_count": self.retry_count,
        }

    @classmethod
    def from_row(cls, row: sqlite3.Row) -> "StepCheckpoint":
        """Create from database row."""
        return cls(
            run_id=row["run_id"],
            step_index=row["step_index"],
            step_name=row["step_name"],
            status=ExecutionStatus(row["status"]),
            started_at=datetime.fromisoformat(row["started_at"]),
            completed_at=datetime.fromisoformat(row["completed_at"])
            if row["completed_at"]
            else None,
            inputs=json.loads(row["inputs"]) if row["inputs"] else None,
            outputs=json.loads(row["outputs"]) if row["outputs"] else None,
            error=row["error"],
            retry_count=row["retry_count"],
        )


class StateStore:
    """
    SQLite-based state store for workflow execution.

    Provides:
    - Execution history tracking
    - Step-level checkpoints for recovery
    - Query capabilities for monitoring
    """

    SCHEMA_VERSION = 1

    def __init__(
        self,
        db_path: str | Path = ".marktoflow/state/workflow-state/state.db",
    ) -> None:
        """
        Initialize the state store.

        Args:
            db_path: Path to SQLite database file
        """
        self.db_path = Path(db_path)
        self.db_path.parent.mkdir(parents=True, exist_ok=True)
        self._init_db()

    @contextmanager
    def _get_connection(self) -> Generator[sqlite3.Connection, None, None]:
        """Get a database connection with row factory."""
        conn = sqlite3.connect(str(self.db_path))
        conn.row_factory = sqlite3.Row
        try:
            yield conn
            conn.commit()
        except Exception:
            conn.rollback()
            raise
        finally:
            conn.close()

    def _init_db(self) -> None:
        """Initialize database schema."""
        with self._get_connection() as conn:
            cursor = conn.cursor()

            # Schema version table
            cursor.execute("""
                CREATE TABLE IF NOT EXISTS schema_version (
                    version INTEGER PRIMARY KEY
                )
            """)

            # Check if we need to create/migrate
            cursor.execute("SELECT version FROM schema_version LIMIT 1")
            row = cursor.fetchone()

            if row is None:
                # Fresh database, create schema
                self._create_schema(cursor)
                cursor.execute(
                    "INSERT INTO schema_version (version) VALUES (?)", (self.SCHEMA_VERSION,)
                )
            elif row["version"] < self.SCHEMA_VERSION:
                # Need migration
                self._migrate_schema(cursor, row["version"])

    def _create_schema(self, cursor: sqlite3.Cursor) -> None:
        """Create database tables."""
        # Executions table
        cursor.execute("""
            CREATE TABLE IF NOT EXISTS executions (
                run_id TEXT PRIMARY KEY,
                workflow_id TEXT NOT NULL,
                workflow_path TEXT NOT NULL,
                status TEXT NOT NULL,
                started_at TEXT NOT NULL,
                completed_at TEXT,
                current_step INTEGER DEFAULT 0,
                total_steps INTEGER DEFAULT 0,
                agent TEXT,
                inputs TEXT,
                outputs TEXT,
                error TEXT,
                metadata TEXT,
                created_at TEXT DEFAULT CURRENT_TIMESTAMP
            )
        """)

        # Checkpoints table
        cursor.execute("""
            CREATE TABLE IF NOT EXISTS checkpoints (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                run_id TEXT NOT NULL,
                step_index INTEGER NOT NULL,
                step_name TEXT NOT NULL,
                status TEXT NOT NULL,
                started_at TEXT NOT NULL,
                completed_at TEXT,
                inputs TEXT,
                outputs TEXT,
                error TEXT,
                retry_count INTEGER DEFAULT 0,
                created_at TEXT DEFAULT CURRENT_TIMESTAMP,
                FOREIGN KEY (run_id) REFERENCES executions(run_id),
                UNIQUE(run_id, step_index)
            )
        """)

        # Indexes for common queries
        cursor.execute(
            "CREATE INDEX IF NOT EXISTS idx_executions_workflow_id ON executions(workflow_id)"
        )
        cursor.execute("CREATE INDEX IF NOT EXISTS idx_executions_status ON executions(status)")
        cursor.execute(
            "CREATE INDEX IF NOT EXISTS idx_executions_started_at ON executions(started_at)"
        )
        cursor.execute("CREATE INDEX IF NOT EXISTS idx_checkpoints_run_id ON checkpoints(run_id)")

    def _migrate_schema(self, cursor: sqlite3.Cursor, from_version: int) -> None:
        """Migrate schema from older version."""
        # Add migration logic here as needed
        logger.info(f"Migrating schema from version {from_version} to {self.SCHEMA_VERSION}")
        cursor.execute("UPDATE schema_version SET version = ?", (self.SCHEMA_VERSION,))

    # Execution management

    def create_execution(self, record: ExecutionRecord) -> None:
        """Create a new execution record."""
        with self._get_connection() as conn:
            cursor = conn.cursor()
            data = record.to_dict()
            cursor.execute(
                """
                INSERT INTO executions (
                    run_id, workflow_id, workflow_path, status, started_at,
                    completed_at, current_step, total_steps, agent, inputs,
                    outputs, error, metadata
                ) VALUES (
                    :run_id, :workflow_id, :workflow_path, :status, :started_at,
                    :completed_at, :current_step, :total_steps, :agent, :inputs,
                    :outputs, :error, :metadata
                )
            """,
                data,
            )

    def update_execution(self, record: ExecutionRecord) -> None:
        """Update an existing execution record."""
        with self._get_connection() as conn:
            cursor = conn.cursor()
            data = record.to_dict()
            cursor.execute(
                """
                UPDATE executions SET
                    status = :status,
                    completed_at = :completed_at,
                    current_step = :current_step,
                    outputs = :outputs,
                    error = :error,
                    metadata = :metadata
                WHERE run_id = :run_id
            """,
                data,
            )

    def get_execution(self, run_id: str) -> ExecutionRecord | None:
        """Get an execution record by run ID."""
        with self._get_connection() as conn:
            cursor = conn.cursor()
            cursor.execute("SELECT * FROM executions WHERE run_id = ?", (run_id,))
            row = cursor.fetchone()
            return ExecutionRecord.from_row(row) if row else None

    def list_executions(
        self,
        workflow_id: str | None = None,
        status: ExecutionStatus | None = None,
        limit: int = 100,
        offset: int = 0,
    ) -> list[ExecutionRecord]:
        """List execution records with optional filters."""
        with self._get_connection() as conn:
            cursor = conn.cursor()

            query = "SELECT * FROM executions WHERE 1=1"
            params: list[Any] = []

            if workflow_id:
                query += " AND workflow_id = ?"
                params.append(workflow_id)

            if status:
                query += " AND status = ?"
                params.append(status.value)

            query += " ORDER BY started_at DESC LIMIT ? OFFSET ?"
            params.extend([limit, offset])

            cursor.execute(query, params)
            return [ExecutionRecord.from_row(row) for row in cursor.fetchall()]

    def get_running_executions(self) -> list[ExecutionRecord]:
        """Get all currently running executions."""
        return self.list_executions(status=ExecutionStatus.RUNNING)

    def get_failed_executions(
        self,
        workflow_id: str | None = None,
        limit: int = 100,
    ) -> list[ExecutionRecord]:
        """Get failed executions for potential retry."""
        return self.list_executions(
            workflow_id=workflow_id,
            status=ExecutionStatus.FAILED,
            limit=limit,
        )

    # Checkpoint management

    def save_checkpoint(self, checkpoint: StepCheckpoint) -> None:
        """Save or update a step checkpoint."""
        with self._get_connection() as conn:
            cursor = conn.cursor()
            data = checkpoint.to_dict()
            cursor.execute(
                """
                INSERT OR REPLACE INTO checkpoints (
                    run_id, step_index, step_name, status, started_at,
                    completed_at, inputs, outputs, error, retry_count
                ) VALUES (
                    :run_id, :step_index, :step_name, :status, :started_at,
                    :completed_at, :inputs, :outputs, :error, :retry_count
                )
            """,
                data,
            )

    def get_checkpoints(self, run_id: str) -> list[StepCheckpoint]:
        """Get all checkpoints for an execution."""
        with self._get_connection() as conn:
            cursor = conn.cursor()
            cursor.execute(
                "SELECT * FROM checkpoints WHERE run_id = ? ORDER BY step_index", (run_id,)
            )
            return [StepCheckpoint.from_row(row) for row in cursor.fetchall()]

    def get_last_checkpoint(self, run_id: str) -> StepCheckpoint | None:
        """Get the last checkpoint for an execution."""
        with self._get_connection() as conn:
            cursor = conn.cursor()
            cursor.execute(
                "SELECT * FROM checkpoints WHERE run_id = ? ORDER BY step_index DESC LIMIT 1",
                (run_id,),
            )
            row = cursor.fetchone()
            return StepCheckpoint.from_row(row) if row else None

    def get_resume_point(self, run_id: str) -> int:
        """
        Get the step index to resume from after a failure.

        Returns:
            Step index to resume from (0 if should restart)
        """
        with self._get_connection() as conn:
            cursor = conn.cursor()

            # Find the last successful step
            cursor.execute(
                """
                SELECT MAX(step_index) as last_success
                FROM checkpoints
                WHERE run_id = ? AND status = 'completed'
            """,
                (run_id,),
            )

            row = cursor.fetchone()
            if row and row["last_success"] is not None:
                # Resume from the next step
                return row["last_success"] + 1

            return 0

    # Cleanup

    def cleanup_old_records(self, days: int = 30) -> int:
        """
        Delete execution records older than specified days.

        Args:
            days: Number of days to retain

        Returns:
            Number of records deleted
        """
        with self._get_connection() as conn:
            cursor = conn.cursor()

            # Calculate cutoff date
            cutoff = datetime.now() - timedelta(days=days)

            # Delete old checkpoints first (foreign key)
            cursor.execute(
                """
                DELETE FROM checkpoints
                WHERE run_id IN (
                    SELECT run_id FROM executions
                    WHERE started_at < ?
                )
            """,
                (cutoff.isoformat(),),
            )

            # Delete old executions
            cursor.execute("DELETE FROM executions WHERE started_at < ?", (cutoff.isoformat(),))

            return cursor.rowcount

    # Statistics

    def get_stats(self, workflow_id: str | None = None) -> dict[str, Any]:
        """Get execution statistics."""
        with self._get_connection() as conn:
            cursor = conn.cursor()

            where_clause = "WHERE workflow_id = ?" if workflow_id else ""
            params = [workflow_id] if workflow_id else []

            # Total executions
            cursor.execute(f"SELECT COUNT(*) as total FROM executions {where_clause}", params)
            total = cursor.fetchone()["total"]

            # By status
            cursor.execute(
                f"""
                SELECT status, COUNT(*) as count
                FROM executions {where_clause}
                GROUP BY status
            """,
                params,
            )
            by_status = {row["status"]: row["count"] for row in cursor.fetchall()}

            # Success rate
            completed = by_status.get("completed", 0)
            failed = by_status.get("failed", 0)
            success_rate = completed / (completed + failed) if (completed + failed) > 0 else 0

            return {
                "total_executions": total,
                "by_status": by_status,
                "success_rate": round(success_rate * 100, 2),
                "completed": completed,
                "failed": failed,
                "running": by_status.get("running", 0),
            }

File: packages/core/src/state.ts
/**
 * State persistence for marktoflow v2.0
 *
 * Provides SQLite-based state storage for workflow checkpoints,
 * execution history, and recovery.
 */

import Database from 'better-sqlite3';
import { WorkflowStatus, StepStatus } from './models.js';

// ============================================================================
// Types
// ============================================================================

export interface ExecutionRecord {
  runId: string;
  workflowId: string;
  workflowPath: string;
  status: WorkflowStatus;
  startedAt: Date;
  completedAt: Date | null;
  currentStep: number;
  totalSteps: number;
  inputs: Record<string, unknown> | null;
  outputs: Record<string, unknown> | null;
  error: string | null;
  metadata: Record<string, unknown> | null;
}

export interface StepCheckpoint {
  runId: string;
  stepIndex: number;
  stepName: string;
  status: StepStatus;
  startedAt: Date;
  completedAt: Date | null;
  inputs: Record<string, unknown> | null;
  outputs: unknown;
  error: string | null;
  retryCount: number;
}

export interface ExecutionStats {
  totalExecutions: number;
  completed: number;
  failed: number;
  running: number;
  successRate: number;
  averageDuration: number | null;
}

// ============================================================================
// StateStore Implementation
// ============================================================================

const SCHEMA_VERSION = 1;

const CREATE_TABLES_SQL = `
  CREATE TABLE IF NOT EXISTS schema_version (
    version INTEGER PRIMARY KEY
  );

  CREATE TABLE IF NOT EXISTS executions (
    run_id TEXT PRIMARY KEY,
    workflow_id TEXT NOT NULL,
    workflow_path TEXT NOT NULL,
    status TEXT NOT NULL,
    started_at TEXT NOT NULL,
    completed_at TEXT,
    current_step INTEGER DEFAULT 0,
    total_steps INTEGER DEFAULT 0,
    inputs TEXT,
    outputs TEXT,
    error TEXT,
    metadata TEXT
  );

  CREATE TABLE IF NOT EXISTS checkpoints (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    run_id TEXT NOT NULL,
    step_index INTEGER NOT NULL,
    step_name TEXT NOT NULL,
    status TEXT NOT NULL,
    started_at TEXT NOT NULL,
    completed_at TEXT,
    inputs TEXT,
    outputs TEXT,
    error TEXT,
    retry_count INTEGER DEFAULT 0,
    FOREIGN KEY (run_id) REFERENCES executions(run_id),
    UNIQUE(run_id, step_index)
  );

  CREATE INDEX IF NOT EXISTS idx_executions_workflow_id ON executions(workflow_id);
  CREATE INDEX IF NOT EXISTS idx_executions_status ON executions(status);
  CREATE INDEX IF NOT EXISTS idx_executions_started_at ON executions(started_at);
  CREATE INDEX IF NOT EXISTS idx_checkpoints_run_id ON checkpoints(run_id);
`;

export class StateStore {
  private db: Database.Database;

  constructor(dbPath: string = '.marktoflow/state/workflow-state.db') {
    this.db = new Database(dbPath);
    this.db.pragma('journal_mode = WAL');
    this.initialize();
  }

  private initialize(): void {
    this.db.exec(CREATE_TABLES_SQL);

    // Check schema version
    const versionRow = this.db.prepare('SELECT version FROM schema_version LIMIT 1').get() as
      | { version: number }
      | undefined;

    if (!versionRow) {
      this.db.prepare('INSERT INTO schema_version (version) VALUES (?)').run(SCHEMA_VERSION);
    }
  }

  // ============================================================================
  // Execution Records
  // ============================================================================

  createExecution(record: ExecutionRecord): void {
    const stmt = this.db.prepare(`
      INSERT INTO executions (
        run_id, workflow_id, workflow_path, status, started_at, completed_at,
        current_step, total_steps, inputs, outputs, error, metadata
      ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    `);

    stmt.run(
      record.runId,
      record.workflowId,
      record.workflowPath,
      record.status,
      record.startedAt.toISOString(),
      record.completedAt?.toISOString() ?? null,
      record.currentStep,
      record.totalSteps,
      record.inputs ? JSON.stringify(record.inputs) : null,
      record.outputs ? JSON.stringify(record.outputs) : null,
      record.error,
      record.metadata ? JSON.stringify(record.metadata) : null
    );
  }

  updateExecution(
    runId: string,
    updates: Partial<Omit<ExecutionRecord, 'runId'>>
  ): void {
    const fields: string[] = [];
    const values: unknown[] = [];

    if (updates.status !== undefined) {
      fields.push('status = ?');
      values.push(updates.status);
    }
    if (updates.completedAt !== undefined) {
      fields.push('completed_at = ?');
      values.push(updates.completedAt?.toISOString() ?? null);
    }
    if (updates.currentStep !== undefined) {
      fields.push('current_step = ?');
      values.push(updates.currentStep);
    }
    if (updates.totalSteps !== undefined) {
      fields.push('total_steps = ?');
      values.push(updates.totalSteps);
    }
    if (updates.outputs !== undefined) {
      fields.push('outputs = ?');
      values.push(updates.outputs ? JSON.stringify(updates.outputs) : null);
    }
    if (updates.error !== undefined) {
      fields.push('error = ?');
      values.push(updates.error);
    }
    if (updates.metadata !== undefined) {
      fields.push('metadata = ?');
      values.push(updates.metadata ? JSON.stringify(updates.metadata) : null);
    }

    if (fields.length === 0) return;

    values.push(runId);
    const sql = `UPDATE executions SET ${fields.join(', ')} WHERE run_id = ?`;
    this.db.prepare(sql).run(...values);
  }

  getExecution(runId: string): ExecutionRecord | null {
    const row = this.db
      .prepare('SELECT * FROM executions WHERE run_id = ?')
      .get(runId) as Record<string, unknown> | undefined;

    return row ? this.rowToExecution(row) : null;
  }

  listExecutions(options: {
    workflowId?: string;
    status?: WorkflowStatus;
    limit?: number;
    offset?: number;
  } = {}): ExecutionRecord[] {
    let sql = 'SELECT * FROM executions WHERE 1=1';
    const params: unknown[] = [];

    if (options.workflowId) {
      sql += ' AND workflow_id = ?';
      params.push(options.workflowId);
    }
    if (options.status) {
      sql += ' AND status = ?';
      params.push(options.status);
    }

    sql += ' ORDER BY started_at DESC';

    if (options.limit) {
      sql += ' LIMIT ?';
      params.push(options.limit);
    }
    if (options.offset) {
      sql += ' OFFSET ?';
      params.push(options.offset);
    }

    const rows = this.db.prepare(sql).all(...params) as Record<string, unknown>[];
    return rows.map((row) => this.rowToExecution(row));
  }

  getRunningExecutions(): ExecutionRecord[] {
    return this.listExecutions({ status: WorkflowStatus.RUNNING });
  }

  getFailedExecutions(limit = 10): ExecutionRecord[] {
    return this.listExecutions({ status: WorkflowStatus.FAILED, limit });
  }

  private rowToExecution(row: Record<string, unknown>): ExecutionRecord {
    return {
      runId: row['run_id'] as string,
      workflowId: row['workflow_id'] as string,
      workflowPath: row['workflow_path'] as string,
      status: row['status'] as WorkflowStatus,
      startedAt: new Date(row['started_at'] as string),
      completedAt: row['completed_at'] ? new Date(row['completed_at'] as string) : null,
      currentStep: row['current_step'] as number,
      totalSteps: row['total_steps'] as number,
      inputs: row['inputs'] ? JSON.parse(row['inputs'] as string) : null,
      outputs: row['outputs'] ? JSON.parse(row['outputs'] as string) : null,
      error: row['error'] as string | null,
      metadata: row['metadata'] ? JSON.parse(row['metadata'] as string) : null,
    };
  }

  // ============================================================================
  // Step Checkpoints
  // ============================================================================

  saveCheckpoint(checkpoint: StepCheckpoint): void {
    const stmt = this.db.prepare(`
      INSERT OR REPLACE INTO checkpoints (
        run_id, step_index, step_name, status, started_at, completed_at,
        inputs, outputs, error, retry_count
      ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    `);

    stmt.run(
      checkpoint.runId,
      checkpoint.stepIndex,
      checkpoint.stepName,
      checkpoint.status,
      checkpoint.startedAt.toISOString(),
      checkpoint.completedAt?.toISOString() ?? null,
      checkpoint.inputs ? JSON.stringify(checkpoint.inputs) : null,
      checkpoint.outputs !== undefined ? JSON.stringify(checkpoint.outputs) : null,
      checkpoint.error,
      checkpoint.retryCount
    );
  }

  getCheckpoints(runId: string): StepCheckpoint[] {
    const rows = this.db
      .prepare('SELECT * FROM checkpoints WHERE run_id = ? ORDER BY step_index')
      .all(runId) as Record<string, unknown>[];

    return rows.map((row) => this.rowToCheckpoint(row));
  }

  getLastCheckpoint(runId: string): StepCheckpoint | null {
    const row = this.db
      .prepare('SELECT * FROM checkpoints WHERE run_id = ? ORDER BY step_index DESC LIMIT 1')
      .get(runId) as Record<string, unknown> | undefined;

    return row ? this.rowToCheckpoint(row) : null;
  }

  getResumePoint(runId: string): number {
    const lastCheckpoint = this.getLastCheckpoint(runId);
    if (!lastCheckpoint) return 0;

    // If last step completed, resume from next step
    if (lastCheckpoint.status === StepStatus.COMPLETED) {
      return lastCheckpoint.stepIndex + 1;
    }

    // Otherwise resume from the failed/pending step
    return lastCheckpoint.stepIndex;
  }

  private rowToCheckpoint(row: Record<string, unknown>): StepCheckpoint {
    return {
      runId: row['run_id'] as string,
      stepIndex: row['step_index'] as number,
      stepName: row['step_name'] as string,
      status: row['status'] as StepStatus,
      startedAt: new Date(row['started_at'] as string),
      completedAt: row['completed_at'] ? new Date(row['completed_at'] as string) : null,
      inputs: row['inputs'] ? JSON.parse(row['inputs'] as string) : null,
      outputs: row['outputs'] ? JSON.parse(row['outputs'] as string) : null,
      error: row['error'] as string | null,
      retryCount: row['retry_count'] as number,
    };
  }

  // ============================================================================
  // Statistics
  // ============================================================================

  getStats(workflowId?: string): ExecutionStats {
    let whereClause = '';
    const params: unknown[] = [];

    if (workflowId) {
      whereClause = 'WHERE workflow_id = ?';
      params.push(workflowId);
    }

    const countRow = this.db
      .prepare(
        `SELECT
          COUNT(*) as total,
          SUM(CASE WHEN status = 'completed' THEN 1 ELSE 0 END) as completed,
          SUM(CASE WHEN status = 'failed' THEN 1 ELSE 0 END) as failed,
          SUM(CASE WHEN status = 'running' THEN 1 ELSE 0 END) as running
        FROM executions ${whereClause}`
      )
      .get(...params) as Record<string, number>;

    const durationRow = this.db
      .prepare(
        `SELECT AVG(
          (julianday(completed_at) - julianday(started_at)) * 86400000
        ) as avg_duration
        FROM executions
        ${whereClause ? whereClause + ' AND' : 'WHERE'} completed_at IS NOT NULL`
      )
      .get(...params) as { avg_duration: number | null };

    const total = countRow['total'] || 0;
    const completed = countRow['completed'] || 0;

    return {
      totalExecutions: total,
      completed,
      failed: countRow['failed'] || 0,
      running: countRow['running'] || 0,
      successRate: total > 0 ? completed / total : 0,
      averageDuration: durationRow.avg_duration,
    };
  }

  // ============================================================================
  // Cleanup
  // ============================================================================

  cleanup(retentionDays: number = 30): number {
    const cutoffDate = new Date();
    cutoffDate.setDate(cutoffDate.getDate() - retentionDays);

    // Delete old checkpoints first (foreign key)
    this.db
      .prepare(
        `DELETE FROM checkpoints WHERE run_id IN (
          SELECT run_id FROM executions WHERE started_at < ?
        )`
      )
      .run(cutoffDate.toISOString());

    // Delete old executions
    const result = this.db
      .prepare('DELETE FROM executions WHERE started_at < ?')
      .run(cutoffDate.toISOString());

    return result.changes;
  }

  close(): void {
    this.db.close();
  }
}
